File: heroic-component/src/main/java/com/spotify/heroic/grammar/IntegerExpression.java
/*
 * Copyright (c) 2015 Spotify AB.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.heroic.grammar;

import com.fasterxml.jackson.annotation.JsonTypeName;
import lombok.Data;

import java.util.concurrent.TimeUnit;

/**
 * int's are represented internally as longs.
 *
 * @author udoprog
 */
@Data
@JsonTypeName("integer")
public final class IntegerExpression implements Expression {
    private final Context context;
    private final long value;

    @Override
    public <R> R visit(final Visitor<R> visitor) {
        return visitor.visitInteger(this);
    }

    @Override
    public IntegerExpression multiply(Expression other) {
        return new IntegerExpression(context.join(other.getContext()),
            value * other.cast(IntegerExpression.class).value);
    }

    @Override
    public IntegerExpression divide(Expression other) {
        return new IntegerExpression(context.join(other.getContext()),
            value / other.cast(IntegerExpression.class).value);
    }

    @Override
    public IntegerExpression sub(Expression other) {
        return new IntegerExpression(context.join(other.getContext()),
            value - other.cast(IntegerExpression.class).value);
    }

    @Override
    public IntegerExpression add(Expression other) {
        return new IntegerExpression(context.join(other.getContext()),
            value + other.cast(IntegerExpression.class).value);
    }

    @Override
    public IntegerExpression negate() {
        return new IntegerExpression(context, -value);
    }

    @SuppressWarnings("unchecked")
    @Override
    public <T extends Expression> T cast(Class<T> to) {
        if (to.isAssignableFrom(IntegerExpression.class)) {
            return (T) this;
        }

        if (to.isAssignableFrom(DurationExpression.class)) {
            return (T) new DurationExpression(context, TimeUnit.MILLISECONDS, value);
        }

        if (to.isAssignableFrom(DoubleExpression.class)) {
            return (T) new DoubleExpression(context, ((Long) value).doubleValue());
        }

        throw context.castError(this, to);
    }

    @Override
    public String toRepr() {
        return Long.toString(value);
    }
}

File: heroic-component/src/main/java/com/spotify/heroic/grammar/IntegerExpression.kt
/*
 * Copyright (c) 2019 Spotify AB.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.heroic.grammar

import java.util.concurrent.TimeUnit

/**
 * int's are represented internally as longs.
 *
 * @author udoprog
 */
data class IntegerExpression(@JvmField val context: Context, val value: Long): Expression {
    override fun getContext() = context

    fun getValueAsInteger(): Int {
        if (value > Integer.MAX_VALUE || value < Integer.MIN_VALUE) {
            throw context.error("cannot be converted to integer")
        }

        return value.toInt()
    }

    override fun <R : Any?> visit(visitor: Expression.Visitor<R>): R {
        return visitor.visitInteger(this)
    }

    override fun multiply(other: Expression): IntegerExpression =
        IntegerExpression(context.join(other.context),
            value * other.cast(IntegerExpression::class.java).value)

    override fun divide(other: Expression): IntegerExpression =
        IntegerExpression(context.join(other.context),
            value / other.cast(IntegerExpression::class.java).value)

    override fun sub(other: Expression): IntegerExpression =
        IntegerExpression(context.join(other.context),
            value - other.cast(IntegerExpression::class.java).value)

    override fun add(other: Expression): IntegerExpression =
        IntegerExpression(context.join(other.context),
            value + other.cast(IntegerExpression::class.java).value)

    override fun negate(): IntegerExpression = IntegerExpression(context, -value)

    @Suppress("UNCHECKED_CAST")
    override fun <T : Expression?> cast(to: Class<T>): T {
        if (to.isAssignableFrom(IntegerExpression::class.java)) {
            return this as T
        }

        if (to.isAssignableFrom(DoubleExpression::class.java)) {
            return DoubleExpression(context, value.toDouble()) as T
        }

        if (to.isAssignableFrom(DurationExpression::class.java)) {
            return DurationExpression(context, TimeUnit.MILLISECONDS, value) as T
        }

        throw context.castError(this, to)
    }

    override fun toRepr() = value.toString()
}